### PR TITLE
change mo_occ dimension detection

### PR DESCRIPTION
**Pull request by tom-bickley**
changed the detection of restricted/unrestricted mo occupancy. Instead of looking at `self.scf_method.mo_occ.ndim` which still returns 1 for unrestricted scf objects, it checks if the first element of `mo_occ` is a `np.float` or `np.ndarray` object.

This seems to work and charged hamiltonians are built but will need to test further.

### CODE REVIEW

File: nbed/ham_builder.py [hunk 1]
```diff
@@ -303,12 +304,24 @@ def _reduce_active_space(
...
+        if isinstance(self.scf_method.mo_occ[0], Number):
             self.occupancy = self.scf_method.mo_occ[active_indices]
-        else:
+        elif isinstance(self.scf_method.mo_occ[0], np.ndarray):
             self.occupancy = np.vstack(
                 (self.scf_method.mo_occ[0], self.scf_method.mo_occ[1])
-            )[:, active_indices]
+                )[:, active_indices]
+        else:
+            raise HamiltonianBuilderError("occupancy dimension error")
```
Comment (MIWdlB) [inline nbed/ham_builder.py:324]: Bit vague, generally a good idea to say which bit of the underlying logic has gone wrong.

```suggestion
            raise HamiltonianBuilderError("Could not determine SCF restriction from MO occupancy.")
```

File: nbed/ham_builder.py [hunk 2]
```diff
@@ -444,8 +457,10 @@ def build(
             core_indices = np.array(core_indices)
             active_indices = np.array(active_indices)
 
+        """
```
Comment (MIWdlB) [inline nbed/ham_builder.py:460]: is unrestricted tapering implemented?

File: nbed/ham_builder.py [hunk 3]
```diff
@@ -303,12 +304,24 @@ def _reduce_active_space(
             )
         ]
 
-        if self.scf_method.mo_occ.ndim == 1:
+        # For some reason the UHF mo_occ is stored as a one-dimensional array of dtype=object, 
+        # where each entry is itself an array. Therefore, ndim=1, when it should be 2. Old code:
+        # if self.scf_method.mo_occ.ndim == 1:
```
Comment (MIWdlB) [inline nbed/ham_builder.py:309]: You can cut out the old code too, my crimes are well enough documented by github 🥲

File: nbed/ham_builder.py (hunk 2 above)
Comment (tom-bickley) [reply nbed/ham_builder.py:460]: I think me and @TimWeaving agreed that symmer would handle this ok but maybe he could confirm?

File: nbed/ham_builder.py (hunk 3 above)
Comment (tom-bickley) [reply nbed/ham_builder.py]: resolved in [5bd6395](https://github.com/UCL-CCS/Nbed/pull/130/commits/5bd6395f834a3dc3faea3741083c293d666619e1)

**Comment by tom-bickley** [comment]
I am going to revert the QMMM functionality from this PR as it is not relevant to the charged Hamiltonian construction issue.

File: nbed/ham_builder.py (hunk 2 above)
Comment (MIWdlB) [reply nbed/ham_builder.py:451]: @tom-bickley @TimWeaving any update?

**Pull request merged**